iDiff/util/tar_utils.go

```diff
@@ -0,0 +1,97 @@
+package tarUtil
+
+import (
+	"archive/tar"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Dir stores a representaiton of a file directory.
+type Dir struct {
+	Name  string
+	Files []string
+	Dirs  []Dir
+}
+
+// UnTar takes in a path to a tar file and writes the untarred version to the provided target.
+// Only untars one level, does not untar nested tars.
+func UnTar(filename string, path string) error {
+	if _, ok := os.Stat(path); ok != nil {
+		os.Mkdir(path, 0777)
+	}
+
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	tr := tar.NewReader(file)
+
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			// end of tar archive
+			break
+		}
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		target := filepath.Join(path, header.Name)
+		switch header.Typeflag {
+
+		// if its a dir and it doesn't exist create it
+		case tar.TypeDir:
+			if _, err := os.Stat(target); err != nil {
+				if err := os.MkdirAll(target, header.FileInfo().Mode()); err != nil {
+					return err
+				}
+				continue
+			}
+
+		// if it's a file create it
+		case tar.TypeReg:
+			currFile, err := os.OpenFile(target, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, header.FileInfo().Mode())
+			if err != nil {
+				return err
+			}
+			defer currFile.Close()
+			_, err = io.Copy(currFile, tr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func isTar(path string) bool {
+	return filepath.Ext(path) == ".tar"
+}
+
+// ExtractTar extracts the tar and any nested tar at the given path.
+// After execution the original tar file is removed and the untarred version is in it place.
+func ExtractTar(path string) error {
+	removeTar := false
+
+	var untarWalkFn func(path string, info os.FileInfo, err error) error
+
+	untarWalkFn = func(path string, info os.FileInfo, err error) error {
```

Comment:
did you try using the `:=` operator, i.e. `untarWalkFn := func()...`? or does that not work because this function is self-referential